Enable specification of output directory

bede opened the pull request:
Follow up to https://twitter.com/beconstant/status/560885880907243520

The script now imports the whole `os` module since `os.makedirs()` is being used.

- [x] Is it mergable?
- [x] Did it pass the tests?
- [x] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage.
- [x] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the ChangeLog?
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

ctb commented:
Hi @bede, thank you for your PR!  Looks like some pep8 violations broke the build; try 'make pep8'.  'make autopep8' will fix them for you if you want to try it.

Also, could you copy/paste in the checklist from here, please?

http://khmer.readthedocs.org/en/v1.3/dev/coding-guidelines-and-review.html#checklist

In addition to a ChangeLog entry, you'll need a test or two for the new functionality.  Should be possible to copy and adapt https://github.com/ged-lab/khmer/blob/master/tests/test_scripts.py#L1471 (test_scripts.py:test_split_paired_reads* functions); you may want to create a new temp directory as per this code, https://github.com/ged-lab/khmer/blob/master/tests/khmer_tst_utils.py#L33, depending on what you're doing.  Let us know if you need any help!

bede commented:
Apologies for not RTFMing @ctb and thank you for the gentle introduction. It's now passing pep8, and I'll look into the tests tomorrow. Impressive build system!

ctb commented:
On Tue, Feb 03, 2015 at 09:38:34AM -0800, Bede Constantinides wrote:
> Apologies for not RTFMing @ctb and thank you for the gentle introduction. It's now passing pep8, and I'll look into the tests tomorrow. Impressive build system!

:) Excellent!

ctb commented:
Also see: https://github.com/ged-lab/khmer/issues/732#issuecomment-71903726.

bede commented:
Shall I add my test as a new function called `def test_split_paired_reads_3`?

ctb commented:
On Mon, Feb 09, 2015 at 02:38:23AM -0800, Bede Constantinides wrote:
> Shall I add my test as a new function called `def test_split_paired_reads_3`?

Sure!

bede commented:
Merged conflicts and rebased into 575df26353c449320b865f88bbfacfd5b6c72050

ctb commented on scripts/split-paired-reads.py:
```diff
@@ -57,8 +61,15 @@ def main():
+    if args.output_directory:
+        if not os.path.exists(args.output_directory):
+            os.makedirs(args.output_directory)
+        out1 = args.output_directory + '/' + os.path.basename(infile) + '.1'
```
maybe use os.path.join here? not sure we'll ever support Windows, admittedly, but.

ctb commented on tests/test_scripts.py:
```diff
@@ -1540,6 +1540,43 @@ def test_split_paired_reads_2_fq():
+
+    # actual output files...
+    outfile1 = utils.get_temp_filename('paired.fq.1')
+    in_dir = os.path.dirname(outfile1)
```
would suggest renaming to output_dir - I think in_dir is confusing, because it's either "put data IN this dir" or "INput dir". Or 'into_dir'.

ctb commented on tests/test_scripts.py:
```diff
@@ -1540,6 +1540,43 @@ def test_split_paired_reads_2_fq():
+    script = scriptpath('split-paired-reads.py')
+    args = [infile, '--output-dir', in_dir]
+
+    utils.runscript(script, args, in_dir)
```
This will work without specifying in_dir, and, more importantly, specifying in_dir means that this test works without specifying --output-dir, because run_script changes to in_dir in this case!

ctb commented:
Nice work - address comments & ping me again, and I should be able to merge!

bede commented:
I think I've addressed your comments @ctb except for using `os.path.join` which fails the test presumably because of the absolute paths it generates... I haven't really investigated why. – is this a sticking point?

ctb commented:
Thx let me take a look!

--
Titus Brown, ctbrown@ucdavis.edu

> On Feb 9, 2015, at 10:20 AM, Bede Constantinides <notifications@github.com> wrote:
> 
> I think I've addressed your comments @ctb except for using os.path.join which fails the test presumably because of the absolute paths it generates... I haven't really investigated why. – is this a sticking point?
> 
> —
> Reply to this email directly or view it on GitHub.
>

bede commented:
Thanks, just let me know if there's anything to fix and I'll rebase again afterwards.

ctb commented:
LGTM - let's see what Jenkins says (tests do pass for me).

ctb commented:
ok to test

bede commented:
Ahah, think I've spotted my mistake

bede commented:
Scratch that last message. Shall I rebase?

ctb commented:
It looks fine to me; one last request: could you add a brief description of -o and --output-dir options to the epilog? See scripts/normalize-by-median.py for how we do the formatting for this.  Then the option info will appear automagically on our docs page.

If you want to do os.path.join and/or rebase, that's fine, but not necessary IMO.

bede commented:
Done! I think?!

ctb commented:
On Mon, Feb 09, 2015 at 09:22:39AM -0800, Bede Constantinides wrote:
> Done! I think?!

heh, can you fix the typo in the ChangeLog? 'outut'

(@mr-c, we need a customer service survey or something... :)

bede commented:
I need to go home!

ctb commented:
it's all good, assuming the tests pass :). thank you!

bede commented:
Thanks for your patience guys!

mr-c commented:
@ctb What sort of survey? I'm missing something.

ctb commented:
"and how was your PR review experience today?"

Mostly a joke...

mr-c commented:
@ctb :-)

@bede, Congratulations on your first commit to the khmer project! Your name will be included in the release notes for the next version and you'll be listed amongst our other contributors in the next software release paper.

bede commented:
Thanks! The main thing is that I wasn't humiliated for naïvely submitting a PR that broke the build (albeit for indenting blank lines). I can see you're fairly relaxed about what you'll merge providing it meets the criteria, which is cool.

maarten1983 commented on scripts/split-paired-reads.py:
```diff
@@ -57,8 +69,15 @@ def main():
     check_space(filenames, args.force)
 
-    out1 = os.path.basename(infile) + '.1'
-    out2 = os.path.basename(infile) + '.2'
```
out3 = os.path.basename(infile) + '.3'

ctb commented:
On Mon, Feb 09, 2015 at 04:06:15PM -0800, Bede Constantinides wrote:
> Thanks! The main thing is that I wasn't humiliated for na??vely submitting a PR that broke the build (albeit for indenting blank lines). I can see you're fairly relaxed about what you'll merge providing it meets the criteria, which is cool. 

Thanks for the feedback and please come again :)
